[wasm] Download Symbols from microsoft symbol server

thaystg opened the pull request:
Creating a draft to download symbols from microsoft symbol server when there is an exception and exceptions are turned on on debugger.

This is a workaround while VS doesn't work on it, which should be the final solution.

If you think that is in the right way, there is still a to do list:
- Read url from config file, or receive as an startup parameter.
- Create a flag to enable and disable this functionality, because this may cause some slowness when creating the callstack of the exception.

The callstack exception without this PR:
![image](https://user-images.githubusercontent.com/4503299/89963541-48661280-dc1e-11ea-9bb2-82cc5c7b3d29.png)


The callstack exception with this PR:
![image](https://user-images.githubusercontent.com/4503299/89963313-a5ad9400-dc1d-11ea-96c6-72d517b3aab4.png)

thaystg committed:
Creating a draft to download symbols from microsoft symbol server when there is an exception and exceptions are turned on on debugger.
This is a workaround while VS doesn't work on it, which should be the final solution.

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +625,80 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+                    Buffer.BlockCopy(data, 4, buffer, 0, 16);
+
+                    var pdbAge = (data[20]
+                        | (data[21] << 8)
```
Give this check and the magic check symbolic meaning or at the very least comments

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +625,80 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+
+                    byte[] result = null;
+
+                    using (WebClient webClient = new WebClient())
```
use httpclient

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +625,80 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
             return true;
         }
 
+        private MethodInfo LoadSymbolsOnDemand(AssemblyInfo asm, uint method_token, SessionId sessionId, CancellationToken token)
```
does this need to be sync? It looks like it should be async

thaystg committed:
Fix what lewing suggested.

thaystg commented:
I'm not sure what is the best way to receive the symbol server url, I was thinking to receive in the same way that we receive browserUri, and pass to DevToolsProxy in run method. If anyone has any better idea, please let me know.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +626,73 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+            for (var i = 0; i < header.Entries.Length; i++)
+            {
+                var entry = header.Entries[i];
+                if (entry.Directory.Type == ImageDebugType.CodeView)
```
You can invert this, and avoid the nesting:

```
if (entry.Directory.Type != ImageDebugType.CodeView)
{
    return null;
}

... rest of the code ..

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +626,73 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+
+                    try
+                    {
+                        using (HttpClient client = new HttpClient())
```
These `usings` can be changed to statements, like:
```
using var client = new HttpClient();
using HttpResponseMessage response = await client.GetAsync(downloadURL);
using Stream streamToReadFrom = await response.Content.ReadAsStreamAsync();
```
.. and avoiding the extra nesting.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +626,73 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+                            using (HttpResponseMessage response = await client.GetAsync(downloadURL))
+                            using (Stream streamToReadFrom = await response.Content.ReadAsStreamAsync())
+                            {
+                                PdbReaderProvider portablePdbReaderProvider = new PdbReaderProvider();
```
nit: `var portablePdbReaderProvider = new PdbReaderProvider()`

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +626,73 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+                    var buffer2 = new byte[(data.Length - 24) - 1];
+                    Buffer.BlockCopy(data, 24, buffer2, 0, (data.Length - 24) - 1);
+                    var pdbName = System.Text.Encoding.UTF8.GetString(buffer2, 0, buffer2.Length);
+                    pdbName = (pdbName.Split(new char[] { '/' }))[pdbName.Split(new char[] { '/' }).Length - 1];
```
Is this a filename? Can you use `Path.GetFileName` on it?

Or maybe at least `pdbName = pdbName.Split(new char[] { '/' }) [^1]`
Or maybe `pdbName = pdbName.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries) [^1]`?

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -540,6 +544,11 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
 
+                        if (method  == null && !asm.Image.HasSymbols)
+                        {
+                            method = await LoadSymbolsOnDemand(asm, method_token, sessionId, token);
```
Do we need a try/catch around this, since it's optional symbols? Maybe even move the one from `LoadSymbolsOnDemand`, to here.

radical reviewed (approved):
Only nitpicks, nothing to block merging. 👍

radical commented:
> I'm not sure what is the best way to receive the symbol server url, I was thinking to receive in the same way that we receive browserUri, and pass to DevToolsProxy in run method. If anyone has any better idea, please let me know.

What does the server url depend on? Does it change?

thaystg committed:
Changing what @radical suggested.

thaystg commented:
> > I'm not sure what is the best way to receive the symbol server url, I was thinking to receive in the same way that we receive browserUri, and pass to DevToolsProxy in run method. If anyone has any better idea, please let me know.
> 
> What does the server url depend on? Does it change?

On VsWindows for example, the user can choose a list of symbol server URL. From Microsoft I think that this is the only one. 

![image](https://user-images.githubusercontent.com/4503299/90405367-043d9c80-e07a-11ea-982c-a3ef2f227974.png)

radical commented:
re:url, you could send a custom message to the proxy to pass the url, and handle that in `MonoProxy`, similar to https://github.com/dotnet/runtime/blob/master/src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs#L340-L341 . We send that one in tests - https://github.com/dotnet/runtime/blob/master/src/mono/wasm/debugger/DebuggerTestSuite/Support.cs#L878 .

How will you get this url from VS though, to pass on to the proxy?

thaystg commented:
The custom message that you suggested is the same that I think we should do.
I have no idea how to get this url from VS, but if we create the custom message, VS team can call this message passing to Proxy the urls.

thaystg committed:
Changed what @radical suggested.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +646,70 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+                        return method;
+                }
+            }
+            return null;
```
Will this mean that we will retry this whole thing the next time we want to get symbols for any method in this assembly? Maybe we need to cache the result of resolving for this asm?

thaystg replied:
No, the next time that we need for symbols for any method the assembly is already populated, and we don't need to do it again.

radical replied:
Sorry, I meant in the case the lookup failed.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +646,70 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+                        | (data[2] << 16)
+                        | (data[3] << 24));
+
+                    if (pdbSignature != 0x53445352) // "SDSR"
```
nit: Is there any documentation/spec/code link that we can provide for the signature/guid extraction stuff here?

thaystg replied:
I copied this from https://github.com/mono/mono/blob/45d996b7ba059ad88d9b3f74182e3a9da16da34d/mono/metadata/debug-mono-ppdb.c#L101

radical replied:
Great. Just a reference to the relative file path, in a comment would be useful.

thaystg replied:
Done! 👍

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -338,6 +347,13 @@ protected override async Task<bool> AcceptCommand(MessageId id, string method, J
+                case "DotnetDebugger.addSymbolServerUrl":
+                    {
+                        string url = args["url"]?.Value<string>();
+                        if (!urlSymbolServerList.Contains(url))
```
```suggestion
                        if (!String.IsNullOrEmpty(url) && !urlSymbolServerList.Contains(url))
```

thaystg committed:
Update src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Update src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
If it's not available on a URL try in the next one in the list.

thaystg committed:
Merge branch 'thays_load_symbols_from_ms_symbol_server' of github.com:thaystg/runtime into thays_load_symbols_from_ms_symbol_server

thaystg commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +646,70 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
             return true;
         }
 
+        async Task<MethodInfo> LoadSymbolsOnDemand(AssemblyInfo asm, uint method_token, SessionId sessionId, CancellationToken token)
```
No you are right, I changed the loop! :)
We should ignore and try the next one.

thaystg commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +646,70 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+                    var downloadURL = urlSymbolServer + "/" + pdbName + "/" + pdbGuid.ToString("N").ToUpper() + pdbAge + "/" + pdbName;
+
+                    
+                    using var client = new HttpClient();
```
done

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +647,77 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+                            SendEvent(sessionId, "Debugger.scriptParsed", scriptSource, token);
+                        }
+                    }
+                    catch (Exception e)
```
And log the exception

thaystg replied:
Done!

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +647,77 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+                        continue;
+                    }
+
+                    if (method != null)
```
nit: move `MethodInfo method;`, and the return inside the try/catch block. And, I guess, log a message if we reach the end of method, because we couldn't find any symbols, or couldn't find any pdb entries(?).

thaystg committed:
Logging error and adding comment about SDSR

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +647,77 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+                        {
+                            var scriptSource = JObject.FromObject(source.ToScriptSource(context.Id, context.AuxData));
+                            SendEvent(sessionId, "Debugger.scriptParsed", scriptSource, token);
+                        }
```
If there was no exception thrown till here, then should we return whatever `method` that we have? or does it make sense to try other symbol servers at this point?

thaystg replied:
No, it doesn't make sense to try other symbol server if we have an exception when trying to send the scriptSource to the browser.
I will change it again.

thaystg replied:
Done! :)

thaystg committed:
Returning if we find the method even if we have an exception sending files to browser.

thaystg committed:
Logging when we don't find the pdb.

thaystg committed:
Changing what @radical suggested.

thaystg committed:
Avoiding that we try to load symbols from the same assembly more than once.

thaystg committed:
Simplifying and adding more log.

radical reviewed (approved):
Thank you for your patience with the review! LGTM! 👍

lewing commented:
> > > I'm not sure what is the best way to receive the symbol server url, I was thinking to receive in the same way that we receive browserUri, and pass to DevToolsProxy in run method. If anyone has any better idea, please let me know.
> > 
> > 
> > What does the server url depend on? Does it change?
> 
> On VsWindows for example, the user can choose a list of symbol server URL. From Microsoft I think that this is the only one.
> 
> ![image](https://user-images.githubusercontent.com/4503299/90405367-043d9c80-e07a-11ea-982c-a3ef2f227974.png)

We should talk to the VS team to find out where these settings are stored and if there is any way to pass them on.

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -617,6 +647,80 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+
+                foreach (var urlSymbolServer in urlSymbolServerList) 
+                {
+                    var downloadURL = urlSymbolServer + "/" + pdbName + "/" + pdbGuid.ToString("N").ToUpper() + pdbAge + "/" + pdbName;
```
nit: 
```suggestion
                    var downloadURL = $"{urlSymbolServer}/{pdbName}/{pdbGuid.ToString("N").ToUpper() + pdbAge}/{pdbName}";
```

thaystg committed:
Update src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs

Co-authored-by: Larry Ewing <lewing@microsoft.com>

thaystg commented:
@lewing , I'm discussing with Diego about how they will pass this info for DebuProxy.

thaystg committed:
Adding support for receive the urlSymbolServer as a parameter in the request as discussed with Diego:

Diego: The response from vs code is that this shouldn't live in js-debug, so I think that passing it around in inspectUri would be the best approach for us

thaystg committed:
Merge remote-tracking branch 'upstream/master' into thays_load_symbols_from_ms_symbol_server

thaystg commented:
@lewing @radical  now this is ready to be merged.
Diego asked to pass the urlSymbolServer as a parameter, like this they don't need to change js-debug. In any case, I let the protocol extension if we need this later and accept the parameter too.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DebuggerProxy.cs:
```diff
@@ -16,9 +17,9 @@ public class DebuggerProxy
         private readonly MonoProxy proxy;
 
-        public DebuggerProxy(ILoggerFactory loggerFactory)
+        public DebuggerProxy(ILoggerFactory loggerFactory, StringValues urlSymbolServerList)
```
Should we maybe take a parameter of type `IEnumerable<string>` instead, or `IList<string>`? And same for `MonoProxy`.

thaystg committed:
Fix @radical suggestion.

radical commented:
Is there anything blocking this now?

thaystg committed:
Removing default symbol server URL, like this we can land this on rc 5.0 without any risk of side effect.

thaystg commented:
I don't think so, I just removed the default microsoft symbol server URL, then we can try to merge it on rc 2.
@lewing can we merge now?

thaystg committed:
Fix compilation.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -11,16 +11,26 @@
+        public MonoProxy(ILoggerFactory loggerFactory, IList<string> urlSymbolServerList, bool hideWebDriver = true) : base(loggerFactory) 
+        { 
+            this.hideWebDriver = hideWebDriver;
+            this.urlSymbolServerList = urlSymbolServerList;
```
Maybe `this.urlSymbolServerList = urlSymbolServerList ?? new List<string>();`?

radical commented on src/mono/wasm/debugger/BrowserDebugHost/Startup.cs:
```diff
@@ -138,7 +139,13 @@ async Task ConnectProxy(HttpContext context)
                             builder => builder.AddConsole().AddFilter(null, LogLevel.Information));
-                        var proxy = new DebuggerProxy(loggerFactory);
+
+                        StringValues urlSymbolServerList;
+                        
+                        context.Request.Query.TryGetValue("urlSymbolServer", out urlSymbolServerList);
+
```
nit:
```suggestion
                        context.Request.Query.TryGetValue("urlSymbolServer", out StringValues urlSymbolServerList);
```

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/TestHarnessStartup.cs:
```diff
@@ -122,7 +123,10 @@ public async Task LaunchAndServe(ProcessStartInfo psi, HttpContext context, Func
                     builder => builder.AddConsole().AddFilter(null, LogLevel.Information));
-                var proxy = new DebuggerProxy(loggerFactory);
+
+                List<String> urlSymbolServerList = new List<String>();
+                
+                var proxy = new DebuggerProxy(loggerFactory, urlSymbolServerList);
```
nit:
```suggestion
                var proxy = new DebuggerProxy(loggerFactory, null);
```

thaystg committed:
Update src/mono/wasm/debugger/DebuggerTestSuite/TestHarnessStartup.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Update src/mono/wasm/debugger/BrowserDebugHost/Startup.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Accepting @radical suggestion.

thaystg committed:
Adding workaround on dotnet/runtime while PR 686 is not merged on Cecil. When it's merged we will not get any side effect with this workaroud.

thaystg commented:
@lewing I removed the default URL of MS Symbol Server, so now, this will not affect the user unless he changes the LaunchSettings.json to use this InspectUri: `"inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}?urlSymbolServer=http://msdl.microsoft.com/download/symbols"`
Then the new implementation will be effective.

thaystg commented:
/backport to release/5.0-rc2
